### PR TITLE
Re-add `payload` to `StoreWorker` to identify the content-type

### PR DESCRIPTION
**Pull request by snazy**
Previously we did not really enforce that the `payload` (a byte value
corresponding to a particular `Content.Type`) was correctly
persisted and as a workaround completely removed the handling of the
`payload` value. This is rather inefficient, because the
`StoreWorker` implementation has to deserialize the whole persisted
value to identify the content type.

Beside the above inefficiency having the right `payload` value is
also a prerequisite to make content-types completely pluggable.

This PR is split into two commits: one with the actual changes and another with the "boring stuff/changes" implied by the actual changes.

Depends on PRs: #4941 #4942

Follow-up PRs: #4944 #4945

See issue #4940

### CODE REVIEW

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java [hunk 1]
```diff
@@ -250,6 +250,15 @@ protected CommitLogEntry commitAttempt(
     Function<Hash, CommitLogEntry> currentCommit =
         h -> h.equals(branchHead) ? currentBranchEntry : null;
 
+    for (KeyWithBytes put : commitParams.getPuts()) {
+      if (put.getPayload() <= (byte) 0) {
```
Comment (dimas-b) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:254]: Do we actually expect this condition to be `true`? We validate payload bytes to be > 0 at the type level, do we not?

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java [hunk 2]
```diff
@@ -632,12 +641,16 @@ protected Stream<Difference> buildDiff(
               if (f.equals(t)) {
                 return null;
               }
+              byte payload =
+                  fromVal != null
+                      ? fromVal.getPayload()
+                      : (toVal != null ? toVal.getPayload() : (byte) 0);
```
Comment (dimas-b) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:647]: Should we check that payloads match if both are set?

File: servers/store/src/main/java/org/projectnessie/server/store/TableCommitMetaStoreWorker.java [hunk 3]
```diff
@@ -246,7 +247,7 @@ public Content applyId(Content content, String id) {
   }
 
   @Override
-  public Content.Type getType(ByteString onRefContent) {
+  public Content.Type getType(byte payload, ByteString onRefContent) {
     ObjectTypes.Content parsed = parse(onRefContent);
```
Comment (dimas-b) [inline servers/store/src/main/java/org/projectnessie/server/store/TableCommitMetaStoreWorker.java:251]: Can we determine type based on `payload` now?

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java (hunk 1 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:254]: We do - probably overly cautious here.

File: servers/store/src/main/java/org/projectnessie/server/store/TableCommitMetaStoreWorker.java (hunk 3 above)
Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/TableCommitMetaStoreWorker.java:251]: **Should** work - but probably not guaranteed for old values.
The follow-up PRs will actually handle the `payload` value and deal with the case when `payload==0`.

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java (hunk 2 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:647]: Actually a different problem.
I'll add a TODO there

Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:647]: #4949

File: versioned/persist/serialize/src/main/java/org/projectnessie/versioned/persist/adapter/serialize/ProtoSerialization.java [hunk 4]
```diff
@@ -197,15 +198,17 @@ public static ContentIdAndBytes protoToContentIdAndBytes(AdapterTypes.ContentIdW
     return ContentIdAndBytes.of(ContentId.of(proto.getContentId().getId()), proto.getValue());
   }
 
-  public static AdapterTypes.ContentIdWithType toProto(ContentId x) {
-    return AdapterTypes.ContentIdWithType.newBuilder()
-        .setContentId(AdapterTypes.ContentId.newBuilder().setId(x.getId()))
-        .setTypeUnused(0)
+  public static AdapterTypes.ContentIdWithPayload toProto(ContentIdAndPayload x) {
```
Comment (dimas-b) [inline versioned/persist/serialize/src/main/java/org/projectnessie/versioned/persist/adapter/serialize/ProtoSerialization.java:201]: This method appears to be called only from test code :thinking:

File: versioned/persist/serialize/src/main/java/org/projectnessie/versioned/persist/adapter/serialize/ProtoSerialization.java [hunk 5]
```diff
@@ -197,15 +198,17 @@ public static ContentIdAndBytes protoToContentIdAndBytes(AdapterTypes.ContentIdW
...
+    return AdapterTypes.ContentIdWithPayload.newBuilder()
+        .setContentId(AdapterTypes.ContentId.newBuilder().setId(x.getContentId().getId()))
+        .setPayload(x.getPayload())
         .build();
   }
 
-  public static ContentId protoToContentId(AdapterTypes.ContentIdWithType proto) {
-    return ContentId.of(proto.getContentId().getId());
+  public static ContentIdAndPayload protoToContentIdAndPayload(
```
Comment (dimas-b) [inline versioned/persist/serialize/src/main/java/org/projectnessie/versioned/persist/adapter/serialize/ProtoSerialization.java:208]: This method appears to be called only from test code :thinking:

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/ContentIdAndPayload.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.adapter;
+
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface ContentIdAndPayload {
```
Comment (dimas-b) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/ContentIdAndPayload.java:21]: Is this class required? I can see usage only in test code :thinking:

**Review by dimas-b — COMMENTED** [review]
LGTM, just on concern about the use of `ContentIdWithPayload`

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/ContentIdAndPayload.java (hunk 6 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/ContentIdAndPayload.java:21]: Weird... could swear, that it's being used.

**Review by dimas-b — APPROVED** [review]

**Pull request merged**